File: lab4-primit/javascript/ErrorsEnum.js
const prompt=require("prompt-sync")({sigint:true}); 

const Error_enumobj = {
	FP_ROUNDING: "FP_ROUNDING",
	FP_OVERFLOW: "FP_OVERFLOW",
	FP_UNDERFLOW: "FP_UNDERFLOW",
	INT_OVERFLOW: "INT_OVERFLOW"
}

const Result_enumobj = {
    A_BIT_DIFFERENT: "A_BIT_DIFFERENT", 
    INFINITY: "INFINITY", 
    ZERO: "ZERO", 
    VERY_DIFFERENT: "VERY_DIFFERENT"
}

// Add a new function that does the opposite of the Error2Result
function result2Error(result) {
	switch (result) {
		case Result_enumobj.A_BIT_DIFFERENT:
			return Error_enumobj.FP_ROUNDING;
			break;
		case Result_enumobj.INFINITY:
			return Error_enumobj.FP_OVERFLOW;
			break;
		case Result_enumobj.ZERO:
			return Error_enumobj.FP_UNDERFLOW;
			break;
		case Result_enumobj.VERY_DIFFERENT:
			return Error_enumobj.INT_OVERFLOW;
			break;
		default:
			return "Invalud Result Value";
	}
}

function error2Result(err){
    switch (err) {
	case Error_enumobj.FP_ROUNDING:
	 return Result_enumobj.A_BIT_DIFFERENT;
	break;
	case Error_enumobj.FP_OVERFLOW:
	    return Result_enumobj.INFINITY;
	break;
	case Error_enumobj.FP_UNDERFLOW:
	    return Result_enumobj.ZERO;
	break;
	case Error_enumobj.INT_OVERFLOW:
	    return Result_enumobj.VERY_DIFFERENT;
	break;
	default:
		return 'Invalid Error value';
}

}

// Change the Object.values to display the results
console.log('Error list: ', Object.values(Result_enumobj));
var validArg = false;
while(!validArg){
    var input = prompt("Input: ");
	// Asign the result variable to the return of the new function
    let result = result2Error(input);
    if (Object.values(Result_enumobj).includes(result)){
        validArg = true;
		console.log(input + " results in " + error2Result(input));
    }
    else{
        console.log(result);
    }
}

File: lab4-primit/java/ErrorsEnum.java
import java.util.Scanner;
import java.util.EnumSet;

public class ErrorsEnum
{
    enum Error { FP_ROUNDING, FP_OVERFLOW, FP_UNDERFLOW, INT_OVERFLOW }

    enum Result { A_BIT_DIFFERENT, INFINITY, ZERO, VERY_DIFFERENT }
    
    private static <E extends Enum<E>> E getEnumElement(String elementTypeName, Class<E> elementType)
    {
        boolean haveResult = false;
        E result = null;
        Scanner stdin = new Scanner(System.in);
        
        while ( ! haveResult )
        {
            System.out.print("Input " + elementTypeName + ": ");
            try
            {
                result = Enum.valueOf(elementType, stdin.next().toUpperCase());
                haveResult = true;
            }
            catch (IllegalArgumentException e)
            {
                System.out.println("Not a valid " + elementTypeName + ".");
                stdin.nextLine(); // skip the invalid input
            }
        }
        
        stdin.close();
        return result;
    }
  
    // Add new function that does the opposite of error2Result
    private static Error result2Error(Result r) {
        Error error = null;
        switch (r) {
            case A_BIT_DIFFERENT:
                error = Error.FP_ROUNDING;
                break;
            case INFINITY:
                error = Error.FP_OVERFLOW;
                break;
            case ZERO:
                error = Error.FP_UNDERFLOW;
                break;
            case VERY_DIFFERENT:
                error = Error.INT_OVERFLOW;
                break;
        }
        return error;
    }

    private static Result error2Result(Error e)
    {
        Result result = null;
        
        switch (e) {
        case FP_ROUNDING:
            result = Result.A_BIT_DIFFERENT;
            break;
        case FP_OVERFLOW:
            result = Result.INFINITY;
            break;
        case FP_UNDERFLOW:
            result = Result.ZERO;
            break;
        case INT_OVERFLOW:
            result = Result.VERY_DIFFERENT;
            break;
        }
        
        return result;
    }

    public static void main(String[] args)
    {
        System.out.print("Known results = ");
        // Ammend for statement to display the results
        for (Result r : EnumSet.allOf(Result.class)) 
        {
            System.out.print(r + " ");
        }
        System.out.println();
        
        // Ammend to get the result
        Result r = getEnumElement("result", Result.class);
        System.out.println(r + " leads to error: " + result2Error(r));
    }
}
